### PR TITLE
Reorganise samples, update samples nuget packages, update build trigger

### PR DESCRIPTION
**Pull request by manvkaur**
1. Reorganised samples into folders for each feature and supported languages underneath
    image
1. Resolved CG vulnerabilities:
 	Upgrade Microsoft.AspNetCore.Http from 2.1.0 to 2.1.22 to fix the vulnerability.
 	Removed Microsoft.AspNetCore.Mvc.Core reference from isolated projects
        Removed STJ reference from in proc project
1. Nuget Package Artifact build - PR trigger removed, path trigger updated.
1. Nuget packages updated in samples C# projects.
1. Some text updates in README files (more README organisation separately)
1. Updated host json to include logging for extension WebJobs
1. Also addressing [https://github.com/Azure/azure-functions-openai-extension/issues/21](https://github.com/Azure/azure-functions-openai-extension/issues/21)
1. Renamed Startup file for Kusto project, removed extra startup file from WebJobs extension
1. Updated packages for extension projects
1. Fixed bug in totalTokens, count was not being set
1. Bumped the version in changelog to 0.9.0

### CODE REVIEW

File: .github/workflows/build.yml
Comment (manvkaur) [inline .github/workflows/build.yml:7]: to allow builds from forks, maintainers need to approve the workflow run after reviewing the PR.

File: build/build-release.yml
Comment (manvkaur) [inline build/build-release.yml:4]: don't want to run this for PR, it's is to create artifacts for nuget

File: samples/chat/csharp-ooproc/CSharpIsolatedSamples.csproj
Comment (manvkaur) [inline samples/chat/csharp-ooproc/CSharpIsolatedSamples.csproj:16]: Upgrade Newtonsoft.Json from 9.0.1 to 13.0.1 to fix the vulnerability.
(Microsoft.AspNetCore.Mvc.Core is the root dependency which doesn't have an update, hence adding this reference for Newtonsoft Json in Isolated).

File: src/WebJobs.Extensions.OpenAI/_CSharpLanguageHelpers.cs
Comment (manvkaur) [inline src/WebJobs.Extensions.OpenAI/_CSharpLanguageHelpers.cs:3]: Legal review recommendation to have this File Header

File: samples/chat/csharp-ooproc/CSharpIsolatedSamples.csproj
Comment (cgillum) [reply samples/chat/csharp-ooproc/CSharpIsolatedSamples.csproj:16]: Interesting. If Microsoft.AspNetCore.Mvc.Core doesn't have an update, then I wonder if we shouldn't be using this package at all since it's apparently not being maintained. Actually, I think @aishwaryabh was thinking of making a change that might remove our dependency on it (as part of an unrelated bug fix)?

Ideally we wouldn't be depending on Newtonsoft.Json in .NET Isolated scenarios.

File: build/build-release.yml
Comment (cgillum) [inline build/build-release.yml:14]: if we're setting `pr: none`, then are these `trigger` settings even needed?

File: samples/chat/csharp-inproc/ChatBotSample.csproj
Comment (cgillum) [inline samples/chat/csharp-inproc/ChatBotSample.csproj:18]: Is this package needed? Usually we just rely on Newtonsoft.Json for .NET in-proc. Using STJ and Newtonsoft together can sometimes cause issues, so I'm wondering if we can just remove STJ from in-proc samples.

File: .github/workflows/build.yml
Comment (manvkaur) [reply .github/workflows/build.yml]: removed this file for now, will test the behavior separately from fork repos

File: samples/textcompletion/csharp-ooproc/TextCompletionIsolated.csproj
Comment (cgillum) [inline samples/textcompletion/csharp-ooproc/TextCompletionIsolated.csproj:16]: As discussed elsewhere, it would be great if we could remove this dependency. Customers generally won't be happy if they see it since OOP is supposed to be standardized on STJ. We can do it in a separate PR if you prefer.

File: samples/textcompletion/csharp-ooproc/host.json
Comment (cgillum) [inline samples/textcompletion/csharp-ooproc/host.json:3]: It might be worth making this host.json file match the in-proc version. It seems like this one just has default settings.

**Review by cgillum — COMMENTED** [review]
Great to see this!

I wonder if we should put semantic search under "search" or "rag" instead of "embeddings". While search does involve "embeddings", I still think of it as a distinct scenario.

One thing I noticed when reviewing these changes is that we need to update the host.json "logging" sections for these samples. Instead of `/logging/logLevel/WebJobs`, it should be `/logging/logLevel/Microsoft.Azure.WebJobs`. This change is necessary to see the logs emitted by the extension. The current config worked when the assembly names started with just "WebJobs" but now that we've changed it to "Microsoft.Azure.WebJobs", we need to make the corresponding change in these host.json files as well.

Should /samples/textcompletion/nodejs/demo.http be moved to /samples/textcompletion?

File: build/build-release.yml
Comment (manvkaur) [reply build/build-release.yml:14]: for CI?

File: samples/chat/csharp-ooproc/ChatBotIsolated.cs
Comment (aishwaryabh) [inline samples/chat/csharp-ooproc/ChatBotIsolated.cs:75]: I can't suggest changes directly to the line so here's the code change example for CreateChatBot:

```
[Function(nameof(CreateChatBot))]
    public static async Task<CreateChatBotOutput> CreateChatBot(
            [HttpTrigger(AuthorizationLevel.Anonymous, "put", Route = "chats/{chatId}")] HttpRequestData req,
            string chatId)
    {
        var responseJson = new { chatId = chatId };

        using StreamReader reader = new StreamReader(req.Body);
        
        string request = await reader.ReadToEndAsync();

        CreateRequest? createRequestBody = JsonSerializer.Deserialize<CreateRequest>(request);

        if (createRequestBody == null)
        {
            throw new ArgumentException("Invalid request body. Make sure that you pass in {\"instructions\": value } as the request body.");
        }

HttpResponseData response = req.CreateResponse(HttpStatusCode.Created);
await response.WriteAsJsonAsync(responseJson);

        return new CreateChatBotOutput
        {
            HttpResponse =response,
            ChatBotCreateRequest = new ChatBotCreateRequest(chatId, createRequestBody.Instructions),
        };
    }

    public class CreateChatBotOutput
    {
        [ChatBotCreateOutput()]
        public ChatBotCreateRequest? ChatBotCreateRequest { get; set; }

        public HttpResponseData? HttpResponse { get; set; }
    }
```

File: build/build-release.yml
Comment (cgillum) [reply build/build-release.yml:14]: How does this ADO pipeline get triggered? I was under the impression that it was manually triggered since you configured `pr: none` above. If it's manually triggered, then why do we need to have path includes? I thought that was only for automatically triggered pipelines?

File: samples/chat/csharp-ooproc/ChatBotIsolated.cs
Comment (aishwaryabh) [reply samples/chat/csharp-ooproc/ChatBotIsolated.cs:75]: This would ensure that we return 201 and we also return the body

Comment (aishwaryabh) [reply samples/chat/csharp-ooproc/ChatBotIsolated.cs:75]: This is the code change for `PostUserResponse`:
```
[Function(nameof(PostUserResponse))]
    public static async Task<PostResponseOutput> PostUserResponse(
        [HttpTrigger(AuthorizationLevel.Anonymous, "post", Route = "chats/{chatId}")] HttpRequestData req,
        string chatId)
    {
        string? userMessage = await req.ReadAsStringAsync();

        if (string.IsNullOrEmpty(userMessage))
        {
            HttpResponseData badResponse = req.CreateResponse(HttpStatusCode.BadRequest);
            await badResponse.WriteStringAsync("Request body is empty");
            return new PostResponseOutput { HttpResponse = badResponse };
        }

        HttpResponseData response = req.CreateResponse(HttpStatusCode.Accepted);

        return new PostResponseOutput
        {
            HttpResponse = response,
            ChatBotPostRequest = new ChatBotPostRequest { UserMessage = userMessage, Id = chatId }
        };
    }

    public class PostResponseOutput
    {
        [ChatBotPostOutput("{chatId}")]
        public ChatBotPostRequest? ChatBotPostRequest { get; set; }

        public HttpResponseData? HttpResponse { get; set; }
    }
````

File: samples/chat/csharp-ooproc/ChatBotIsolated.cs
Comment (aishwaryabh) [inline samples/chat/csharp-ooproc/ChatBotIsolated.cs:77]: Can we also get this change in for `GetChatState`? That way we return `application/json` content type for both in proc and OOP:
```
[Function(nameof(GetChatState))]
public static async Task<HttpResponseData> GetChatState(
   [HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = "chats/{chatId}")] HttpRequestData req,
   string chatId,
   [ChatBotQueryInput("{chatId}", TimestampUtc = "{Query.timestampUTC}")] ChatBotState state,
   FunctionContext context)
{
    var response = req.CreateResponse(HttpStatusCode.OK);
    await response.WriteAsJsonAsync(state);
    return response;
}
```

Comment (aishwaryabh) [reply samples/chat/csharp-ooproc/ChatBotIsolated.cs:77]: We would also need to update `ChatBotState.cs ` and `ChatMessage.cs` within the worker such that each property also has `JsonPropertyName` as a header.

Example:
```
public class ChatBotState
{
    /// <summary>
    /// Gets the ID of the chat bot.
    /// </summary>
    [JsonPropertyName("id")]
    public string? Id { get; set; }

    /// <summary>
    /// Gets if chat bot exists.
    /// </summary>
    [JsonPropertyName("exists")]
    public bool Exists { get; set; }

    /// <summary>
    /// Gets status of chat bot. Options are Uninitialzied, Active, or Expired.
    /// </summary>
    [JsonPropertyName("status")]
    public string? Status { get; set; }

    /// <summary>
    /// Gets timestamp of when chat bot is created.
    /// </summary>
    [JsonPropertyName("createdAt")]
    public DateTime CreatedAt { get; set; }

    /// <summary>
    /// Gets timestamp of when chat bot is last updated.
    /// </summary>
    [JsonPropertyName("lastUpdatedAt")]
    public DateTime LastUpdatedAt { get; set; }

    /// <summary>
    /// Gets number of total messages for chat bot.
    /// </summary>
    [JsonPropertyName("totalMessages")]
    public int TotalMessages { get; set; }

    /// <summary>
    /// Gets number of total tokens for chatbot.
    /// </summary>
    [JsonPropertyName("totalTokens")]
    public int TotalTokens { get; set; }

    /// <summary>
    /// Gets a list of the recent messages from the chatbot.
    /// </summary>
    [JsonPropertyName("recentMessages")]
    public IReadOnlyList<ChatMessage> RecentMessages { get; set; } = Array.Empty<ChatMessage>();
}
```

Comment (aishwaryabh) [reply samples/chat/csharp-ooproc/ChatBotIsolated.cs:77]: Changes for ChatMessage:
```
public class ChatMessage
{
    /// <summary>
    /// Initializes a new instance of the <see cref="ChatMessage"/> class.
    /// </summary>
    /// <param name="content">The content of the message.</param>
    /// <param name="role">The role of the chat agent.</param>
    public ChatMessage(string content, string role)
    {
        this.Content = content;
        this.Role = role;
    }

    /// <summary>
    /// Gets or sets the content of the message.
    /// </summary>
    [JsonPropertyName("content")]
    public string Content { get; set; }

    /// <summary>
    /// Gets or sets the role of the chat agent.
    /// </summary>
    [JsonPropertyName("role")]
    public string Role { get; set; }
}
```

Comment (aishwaryabh) [reply samples/chat/csharp-ooproc/ChatBotIsolated.cs:77]: This ensures the E2E test works for dotnet isolated as well to make sure both samples are returning the same status codes and body content as expected

Comment (cgillum) [reply samples/chat/csharp-ooproc/ChatBotIsolated.cs:77]: @aishwaryabh if you want, you can alternatively create a PR on top of this one to add the updated samples directly. :)

Comment (manvkaur) [reply samples/chat/csharp-ooproc/ChatBotIsolated.cs:84]: made these changes in worker and samples, wondering if we should publish new package or wait for significant changes.

File: build/build-release.yml
Comment (manvkaur) [reply build/build-release.yml:14]: ADO has bit different trigger settings than github, the keyword trigger is used for continuous integration triggers (here after merge to main branch and from included file paths only)

File: src/WebJobs.Extensions.OpenAI/Agents/ChatBotService.cs
Comment (manvkaur) [inline src/WebJobs.Extensions.OpenAI/Agents/ChatBotService.cs:275]: missed during migration to table storage, added the validation to tests to avoid in future

File: samples/textcompletion/csharp-ooproc/TextCompletions.cs
Comment (manvkaur) [inline samples/textcompletion/csharp-ooproc/TextCompletions.cs:33]: @cgillum , I had confusion with Generic completion, how can I create HttpResponseData object below since we do not pass the HttpRequestData object.

File: samples/chat/csharp-ooproc/ChatBotIsolated.cs
Comment (aishwaryabh) [reply samples/chat/csharp-ooproc/ChatBotIsolated.cs:84]: I think we can wait; these samples should still work without the worker changes. The worker changes just ensure the E2E test works as expected (since the JSON property gets bound as pascal casing instead of camel casing)

File: samples/textcompletion/csharp-ooproc/TextCompletions.cs
Comment (cgillum) [reply samples/textcompletion/csharp-ooproc/TextCompletions.cs:33]: There are two options we can consider:
1. Change the function signature to take an `HttpRequestData` instead of a `PromptPayload` or
2. Add a `FunctionsContext` parameter to the function and then use it to manually create an `HttpResponseData` object via its constructor - something like `new HttpResponseData(context);`

I'm not sure which option is the best but it may depend on the behavior of the binding expression rules. The second option might be the safest since it doesn't mess with the input binding behavior at all.

File: samples/chat/csharp-ooproc/ChatBotIsolated.cs
Comment (manvkaur) [reply samples/chat/csharp-ooproc/ChatBotIsolated.cs:84]: done

File: samples/textcompletion/csharp-ooproc/TextCompletions.cs
Comment (manvkaur) [reply samples/textcompletion/csharp-ooproc/TextCompletions.cs:33]: used option 1

File: samples/rag/csharp-inproc/SemanticSearchEmbeddings.csproj
Comment (cgillum) [inline samples/rag/csharp-inproc/SemanticSearchEmbeddings.csproj:7]: We should update this to 4.3.0

**Review by cgillum — APPROVED** [review]
LGTM!

File: samples/rag/csharp-inproc/SemanticSearchEmbeddings.csproj
Comment (manvkaur) [reply samples/rag/csharp-inproc/SemanticSearchEmbeddings.csproj]: done

**Pull request merged**
